File: packages/go/qorechain/pqc/pqc_test.go
package pqc

import (
	"bytes"
	"testing"

	pqcv1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/pqc/v1"
)

func TestKeypairSizes(t *testing.T) {
	kp, err := GeneratePQCKeypair()
	if err != nil {
		t.Fatal(err)
	}
	if len(kp.PublicKey) != MLDSA87PublicKeyLength {
		t.Errorf("public key len = %d, want %d", len(kp.PublicKey), MLDSA87PublicKeyLength)
	}
	if len(kp.SecretKey) != MLDSA87SecretKeyLength {
		t.Errorf("secret key len = %d, want %d", len(kp.SecretKey), MLDSA87SecretKeyLength)
	}
	if MLDSA87PublicKeyLength != 2592 || MLDSA87SecretKeyLength != 4896 || MLDSA87SignatureLength != 4627 {
		t.Errorf("constants drifted: %d/%d/%d", MLDSA87PublicKeyLength, MLDSA87SecretKeyLength, MLDSA87SignatureLength)
	}
}

func TestSignVerify(t *testing.T) {
	kp, err := GeneratePQCKeypair()
	if err != nil {
		t.Fatal(err)
	}
	msg := []byte("qorechain hybrid signature test")
	sig, err := PQCSign(kp.SecretKey, msg)
	if err != nil {
		t.Fatal(err)
	}
	if len(sig) != MLDSA87SignatureLength {
		t.Errorf("signature len = %d, want %d", len(sig), MLDSA87SignatureLength)
	}
	if !PQCVerify(kp.PublicKey, msg, sig) {
		t.Error("valid signature should verify")
	}
}

func TestTamperFails(t *testing.T) {
	kp, _ := GeneratePQCKeypair()
	msg := []byte("hello")
	sig, _ := PQCSign(kp.SecretKey, msg)

	// Tampered message.
	if PQCVerify(kp.PublicKey, []byte("hellp"), sig) {
		t.Error("tampered message should not verify")
	}
	// Tampered signature.
	bad := append([]byte(nil), sig...)
	bad[0] ^= 0xff
	if PQCVerify(kp.PublicKey, msg, bad) {
		t.Error("tampered signature should not verify")
	}
}

func TestAlgorithmHelpers(t *testing.T) {
	if AlgorithmDilithium5 != 1 || AlgorithmMLKEM1024 != 2 {
		t.Errorf("algorithm constants drifted: %d %d", AlgorithmDilithium5, AlgorithmMLKEM1024)
	}
	if AlgorithmName(AlgorithmDilithium5) != "dilithium5" {
		t.Errorf("name = %q", AlgorithmName(AlgorithmDilithium5))
	}
	if AlgorithmName(AlgorithmMLKEM1024) != "mlkem1024" {
		t.Errorf("name = %q", AlgorithmName(AlgorithmMLKEM1024))
	}
	if !IsSignatureAlgorithm(AlgorithmDilithium5) {
		t.Error("dilithium5 should be a signature algorithm")
	}
	if IsSignatureAlgorithm(AlgorithmMLKEM1024) {
		t.Error("mlkem1024 is not a signature algorithm")
	}
	if HybridSigTypeURL != "/qorechain.pqc.v1.PQCHybridSignature" {
		t.Errorf("type url = %q", HybridSigTypeURL)
	}
}

func TestEncodeHybridSignatureExtensionProto(t *testing.T) {
	kp, _ := GeneratePQCKeypair()
	sig, _ := PQCSign(kp.SecretKey, []byte("m"))

	value, err := EncodeHybridSignatureExtension(AlgorithmDilithium5, sig, kp.PublicKey)
	if err != nil {
		t.Fatal(err)
	}

	// CONSENSUS-CRITICAL: the extension value must be protobuf, not Go-JSON. Its
	// first byte is the field-1 varint tag 0x08, NEVER '{' (0x7b) — the chain's
	// tx decoder rejects a 0x7b-leading value (misread as field 15 start_group).
	if len(value) == 0 {
		t.Fatal("encoded extension value is empty")
	}
	if value[0] != 0x08 {
		t.Errorf("extension value[0] = 0x%02x, want 0x08 (proto field-1 tag)", value[0])
	}
	if value[0] == 0x7b {
		t.Error("extension value is Go-JSON ('{' = 0x7b); it must be protobuf")
	}

	// Round-trip: unmarshal into the generated PQCHybridSignature message.
	var decoded pqcv1.PQCHybridSignature
	if err := decoded.Unmarshal(value); err != nil {
		t.Fatalf("proto unmarshal failed: %v", err)
	}
	if int(decoded.AlgorithmID) != AlgorithmDilithium5 {
		t.Errorf("AlgorithmID = %d, want %d", decoded.AlgorithmID, AlgorithmDilithium5)
	}
	if !bytes.Equal(decoded.PQCSignature, sig) {
		t.Error("PqcSignature did not round-trip")
	}
	if !bytes.Equal(decoded.PQCPublicKey, kp.PublicKey) {
		t.Error("PqcPublicKey did not round-trip")
	}
}

func TestEncodeHybridSignatureExtensionOmitsPublicKey(t *testing.T) {
	kp, _ := GeneratePQCKeypair()
	sig, _ := PQCSign(kp.SecretKey, []byte("m"))
	value, err := EncodeHybridSignatureExtension(AlgorithmDilithium5, sig, nil)
	if err != nil {
		t.Fatal(err)
	}
	if len(value) == 0 || value[0] != 0x08 {
		t.Fatalf("extension value must be proto-encoded starting with 0x08")
	}
	var decoded pqcv1.PQCHybridSignature
	if err := decoded.Unmarshal(value); err != nil {
		t.Fatalf("proto unmarshal failed: %v", err)
	}
	if len(decoded.PQCPublicKey) != 0 {
		t.Error("PqcPublicKey should be empty when public key is nil")
	}
}

func TestEncodeHybridSignatureExtensionValidation(t *testing.T) {
	kp, _ := GeneratePQCKeypair()
	sig, _ := PQCSign(kp.SecretKey, []byte("m"))

	// Non-signature algorithm.
	if _, err := EncodeHybridSignatureExtension(AlgorithmMLKEM1024, sig, nil); err == nil {
		t.Error("expected error for non-signature algorithm")
	}
	// Empty signature.
	if _, err := EncodeHybridSignatureExtension(AlgorithmDilithium5, nil, nil); err == nil {
		t.Error("expected error for empty signature")
	}
	// Wrong signature length.
	if _, err := EncodeHybridSignatureExtension(AlgorithmDilithium5, sig[:10], nil); err == nil {
		t.Error("expected error for wrong signature length")
	}
	// Wrong public key length.
	if _, err := EncodeHybridSignatureExtension(AlgorithmDilithium5, sig, []byte{1, 2, 3}); err == nil {
		t.Error("expected error for wrong public key length")
	}
}

// TestPQCVerifyRejectsWrongSignatureLength is a regression test for a real
// cross-binding malleability: the underlying CIRCL implementation verifies a
// 4627-byte signature with one extra trailing byte as TRUE, while the Rust
// fips204 binding rejects it — so the same bytes were valid to the Go SDK and
// invalid to the Rust SDK. PQCVerify must enforce the exact size itself.
func TestPQCVerifyRejectsWrongSignatureLength(t *testing.T) {
	kp, err := GeneratePQCKeypair()
	if err != nil {
		t.Fatal(err)
	}
	msg := []byte("length strictness is consensus-critical")
	sig, err := PQCSign(kp.SecretKey, msg)
	if err != nil {
		t.Fatal(err)
	}
	if len(sig) != MLDSA87SignatureLength {
		t.Fatalf("signature len = %d, want %d", len(sig), MLDSA87SignatureLength)
	}

	// The exact-length signature must still verify.
	if !PQCVerify(kp.PublicKey, msg, sig) {
		t.Fatal("a valid exact-length signature must verify")
	}

	// One extra trailing byte — the case the library accepts on its own.
	padded := make([]byte, 0, len(sig)+1)
	padded = append(padded, sig...)
	padded = append(padded, 0x00)
	if PQCVerify(kp.PublicKey, msg, padded) {
		t.Errorf("signature of %d bytes (one extra trailing byte) must be rejected", len(padded))
	}

	// Truncated by one byte.
	if PQCVerify(kp.PublicKey, msg, sig[:len(sig)-1]) {
		t.Errorf("signature of %d bytes (truncated) must be rejected", len(sig)-1)
	}

	// Empty and nil signatures.
	if PQCVerify(kp.PublicKey, msg, nil) || PQCVerify(kp.PublicKey, msg, []byte{}) {
		t.Error("an empty signature must be rejected")
	}
}

// TestPQCVerifyRejectsWrongPublicKeyLength asserts the public key is held to its
// exact size too, on both sides of the correct length.
func TestPQCVerifyRejectsWrongPublicKeyLength(t *testing.T) {
	kp, err := GeneratePQCKeypair()
	if err != nil {
		t.Fatal(err)
	}
	msg := []byte("public keys are exact-sized")
	sig, err := PQCSign(kp.SecretKey, msg)
	if err != nil {
		t.Fatal(err)
	}

	padded := make([]byte, 0, len(kp.PublicKey)+1)
	padded = append(padded, kp.PublicKey...)
	padded = append(padded, 0x00)

	for name, pub := range map[string][]byte{
		"one extra trailing byte": padded,
		"truncated":               kp.PublicKey[:len(kp.PublicKey)-1],
		"empty":                   {},
		"nil":                     nil,
	} {
		if PQCVerify(pub, msg, sig) {
			t.Errorf("public key (%s, %d bytes) must be rejected", name, len(pub))
		}
	}
}

// TestPQCSignRejectsWrongSecretKeyLength asserts signing refuses a secret key
// that is not exactly MLDSA87SecretKeyLength bytes, before the library sees it.
func TestPQCSignRejectsWrongSecretKeyLength(t *testing.T) {
	kp, err := GeneratePQCKeypair()
	if err != nil {
		t.Fatal(err)
	}
	padded := make([]byte, 0, len(kp.SecretKey)+1)
	padded = append(padded, kp.SecretKey...)
	padded = append(padded, 0x00)

	for name, sec := range map[string][]byte{
		"one extra trailing byte": padded,
		"truncated":               kp.SecretKey[:len(kp.SecretKey)-1],
		"empty":                   {},
		"nil":                     nil,
	} {
		if _, err := PQCSign(sec, []byte("m")); err == nil {
			t.Errorf("secret key (%s, %d bytes) must be rejected", name, len(sec))
		}
	}
	// The exact-length key still signs.
	if _, err := PQCSign(kp.SecretKey, []byte("m")); err != nil {
		t.Fatalf("exact-length secret key must sign: %v", err)
	}
}

File: packages/java/src/test/java/io/github/qorechain/PqcTest.java
package io.github.qorechain;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.github.qorechain.pqc.HybridSignatureExtension;
import io.github.qorechain.pqc.Pqc;
import io.github.qorechain.pqc.PqcAlgorithm;
import io.github.qorechain.pqc.PqcKeypair;
import io.github.qorechain.tx.HybridTx;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import org.junit.jupiter.api.Test;

/**
 * ML-DSA-87 sizes, sign/verify/tamper, deterministic keygen, and protobuf encoding
 * of the hybrid-signature extension.
 */
class PqcTest {

    @Test
    void mlDsa87Sizes() {
        assertEquals(2592, Pqc.ML_DSA_87_PUBLIC_KEY_LENGTH);
        assertEquals(4896, Pqc.ML_DSA_87_SECRET_KEY_LENGTH);
        assertEquals(4627, Pqc.ML_DSA_87_SIGNATURE_LENGTH);
        PqcKeypair kp = Pqc.generatePqcKeypair();
        assertEquals(2592, kp.publicKey.length);
        assertEquals(4896, kp.secretKey.length);
        byte[] sig = Pqc.pqcSign(kp.secretKey, "transaction body bytes".getBytes(StandardCharsets.UTF_8));
        assertEquals(4627, sig.length);
    }

    @Test
    void signVerifyAndTamper() {
        PqcKeypair kp = Pqc.generatePqcKeypair();
        byte[] msg = "transaction body bytes".getBytes(StandardCharsets.UTF_8);
        byte[] sig = Pqc.pqcSign(kp.secretKey, msg);
        assertTrue(Pqc.pqcVerify(kp.publicKey, msg, sig));
        byte[] tampered = "transaction body bytesX".getBytes(StandardCharsets.UTF_8);
        assertFalse(Pqc.pqcVerify(kp.publicKey, tampered, sig));
    }

    @Test
    void deterministicKeygenFromSeed() {
        byte[] seed = new byte[32];
        Arrays.fill(seed, (byte) 9);
        PqcKeypair k1 = Pqc.generatePqcKeypair(seed);
        PqcKeypair k2 = Pqc.generatePqcKeypair(seed);
        assertArrayEquals(k1.publicKey, k2.publicKey);
        assertArrayEquals(k1.secretKey, k2.secretKey);
    }

    @Test
    void invalidSeedLengthThrows() {
        assertThrows(IllegalArgumentException.class, () -> Pqc.generatePqcKeypair(new byte[10]));
    }

    @Test
    void hybridSigTypeUrl() {
        assertEquals("/qorechain.pqc.v1.PQCHybridSignature", Pqc.HYBRID_SIG_TYPE_URL);
    }

    @Test
    void extensionProtoOmitsPublicKeyWhenAbsent() throws Exception {
        PqcKeypair kp = Pqc.generatePqcKeypair();
        byte[] sig = Pqc.pqcSign(kp.secretKey, "m".getBytes(StandardCharsets.UTF_8));
        HybridSignatureExtension ext =
                Pqc.buildHybridSignatureExtension(PqcAlgorithm.ALGORITHM_DILITHIUM5, sig, null);
        com.google.protobuf.Any any = HybridTx.encodeHybridExtension(ext);
        // Protobuf-encoded: leading field-1 varint tag 0x08, NEVER JSON 0x7b ('{').
        assertEquals((byte) 0x08, any.getValue().byteAt(0));
        assertNotEquals((byte) 0x7b, any.getValue().byteAt(0));
        qorechain.pqc.v1.Hybrid.PQCHybridSignature decoded =
                qorechain.pqc.v1.Hybrid.PQCHybridSignature.parseFrom(any.getValue());
        assertEquals(1, decoded.getAlgorithmId());
        assertArrayEquals(sig, decoded.getPqcSignature().toByteArray());
        assertEquals(0, decoded.getPqcPublicKey().size());
    }

    @Test
    void extensionProtoIncludesPublicKeyWhenPresent() throws Exception {
        PqcKeypair kp = Pqc.generatePqcKeypair();
        byte[] sig = Pqc.pqcSign(kp.secretKey, "m".getBytes(StandardCharsets.UTF_8));
        HybridSignatureExtension ext =
                Pqc.buildHybridSignatureExtension(
                        PqcAlgorithm.ALGORITHM_DILITHIUM5, sig, kp.publicKey);
        com.google.protobuf.Any any = HybridTx.encodeHybridExtension(ext);
        assertEquals((byte) 0x08, any.getValue().byteAt(0));
        qorechain.pqc.v1.Hybrid.PQCHybridSignature decoded =
                qorechain.pqc.v1.Hybrid.PQCHybridSignature.parseFrom(any.getValue());
        assertEquals(1, decoded.getAlgorithmId());
        assertArrayEquals(kp.publicKey, decoded.getPqcPublicKey().toByteArray());
        assertEquals(1, ext.algorithmId);
    }

    @Test
    void buildExtensionRejectsEmptyAndWrongSize() {
        assertThrows(
                IllegalArgumentException.class,
                () -> Pqc.buildHybridSignatureExtension(
                        PqcAlgorithm.ALGORITHM_DILITHIUM5, new byte[0], null));
        assertThrows(
                IllegalArgumentException.class,
                () -> Pqc.buildHybridSignatureExtension(
                        PqcAlgorithm.ALGORITHM_DILITHIUM5, new byte[10], null));
    }

    /**
     * The FIPS-204 encodings are fixed-length, and the five official QoreChain
     * bindings must agree on that: an ML-DSA-87 signature is 4627 bytes, not 4628
     * with a byte of trailing garbage. (A probe found the Go binding accepting the
     * over-long form while Rust rejected it — a disagreement the chain cannot
     * afford.) Verification therefore rejects on length, before the library runs.
     */
    @Test
    void verifyRejectsSignaturesThatAreNotExactlySpecLength() {
        PqcKeypair kp = Pqc.generatePqcKeypair();
        byte[] msg = "transaction body bytes".getBytes(StandardCharsets.UTF_8);
        byte[] sig = Pqc.pqcSign(kp.secretKey, msg);
        assertEquals(Pqc.ML_DSA_87_SIGNATURE_LENGTH, sig.length);
        // The exact-length signature still verifies.
        assertTrue(Pqc.pqcVerify(kp.publicKey, msg, sig));

        // One extra trailing byte: same 4627-byte prefix, still rejected.
        byte[] tooLong = Arrays.copyOf(sig, sig.length + 1);
        assertEquals(Pqc.ML_DSA_87_SIGNATURE_LENGTH + 1, tooLong.length);
        assertFalse(Pqc.pqcVerify(kp.publicKey, msg, tooLong));

        // Truncated by one byte: rejected.
        byte[] truncated = Arrays.copyOf(sig, sig.length - 1);
        assertFalse(Pqc.pqcVerify(kp.publicKey, msg, truncated));

        // Null / empty are rejected as failures, not crashes.
        assertFalse(Pqc.pqcVerify(kp.publicKey, msg, null));
        assertFalse(Pqc.pqcVerify(kp.publicKey, msg, new byte[0]));
    }

    @Test
    void verifyRejectsPublicKeysThatAreNotExactlySpecLength() {
        PqcKeypair kp = Pqc.generatePqcKeypair();
        byte[] msg = "transaction body bytes".getBytes(StandardCharsets.UTF_8);
        byte[] sig = Pqc.pqcSign(kp.secretKey, msg);

        assertFalse(Pqc.pqcVerify(Arrays.copyOf(kp.publicKey, kp.publicKey.length + 1), msg, sig));
        assertFalse(Pqc.pqcVerify(Arrays.copyOf(kp.publicKey, kp.publicKey.length - 1), msg, sig));
        assertFalse(Pqc.pqcVerify(null, msg, sig));
    }

    @Test
    void signRejectsSecretKeysThatAreNotExactlySpecLength() {
        PqcKeypair kp = Pqc.generatePqcKeypair();
        byte[] msg = "m".getBytes(StandardCharsets.UTF_8);
        byte[] tooLong = Arrays.copyOf(kp.secretKey, kp.secretKey.length + 1);
        byte[] truncated = Arrays.copyOf(kp.secretKey, kp.secretKey.length - 1);

        assertThrows(IllegalArgumentException.class, () -> Pqc.pqcSign(tooLong, msg));
        assertThrows(IllegalArgumentException.class, () -> Pqc.pqcSign(truncated, msg));
        assertThrows(IllegalArgumentException.class, () -> Pqc.pqcSign(null, msg));
        assertThrows(IllegalArgumentException.class, () -> Pqc.pqcSignHedged(tooLong, msg));
    }

    @Test
    void algorithmConstants() {
        assertEquals(1, PqcAlgorithm.ALGORITHM_DILITHIUM5);
        assertEquals(2, PqcAlgorithm.ALGORITHM_MLKEM1024);
        assertTrue(PqcAlgorithm.isSignatureAlgorithm(PqcAlgorithm.ALGORITHM_DILITHIUM5));
        assertFalse(PqcAlgorithm.isSignatureAlgorithm(PqcAlgorithm.ALGORITHM_MLKEM1024));
    }
}
